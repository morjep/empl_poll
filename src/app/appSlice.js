import { createSlice, createAsyncThunk } from "@reduxjs/toolkit";
import { _getQuestions, _saveQuestion, _getUsers, _saveQuestionAnswer } from "../utils/_DATA";

const initialState = {
  questions: {},
  users: {},
  authedUser: null,
  authStatus: null,
  statusUsersAPI: "idle",
  statusQuestionsAPI: "idle",
  error: null,
};

const appSlice = createSlice({
  name: "app",
  initialState,
  reducers: {
    logout: (state) => {
      state.authedUser = null;
      state.authStatus = null;
    },
    login: (state, action) => {
      const { name, pwd } = action.payload;
      const user = state.users[name];
      if (user && user.password === pwd) {
        state.authedUser = user.id;
        state.authStatus = "User authenticated";
      }
      if (state.authedUser === null) {
        state.authStatus = "User not found";
      }
    },
  },
  extraReducers(builder) {
    builder
      .addCase(fetchQuestions.pending, (state, action) => {
        state.statusQuestionsAPI = "loading";
      })
      .addCase(fetchQuestions.fulfilled, (state, action) => {
        state.statusQuestionsAPI = "succeeded";
        state.questions = action.payload;
      })
      .addCase(fetchQuestions.rejected, (state, action) => {
        state.statusQuestionsAPI = "failed";
        state.error = action.error.message;
      })
      .addCase(saveNewQuestion.fulfilled, (state, action) => {
        if (action.payload) {
          const question = action.payload;
          state.questions[question.id] = question;
          state.users[state.authedUser].questions.push(question.id);
        }
      })
      .addCase(fetchUsers.pending, (state, action) => {
        state.statusUsersAPI = "loading";
      })
      .addCase(fetchUsers.fulfilled, (state, action) => {
        state.statusUsersAPI = "succeeded";
        state.users = action.payload;
      })
      .addCase(fetchUsers.rejected, (state, action) => {
        state.statusUsersAPI = "failed";
        state.error = action.error.message;
      })
      .addCase(saveUserAnswer.fulfilled, (state, action) => {
        if (action) {
          const { authedUser, qid, answer } = action.payload;
          state.users[authedUser].answers[qid] = answer;

          if (answer === "optionOne") {
            state.questions[qid].optionOne.votes.push(authedUser);
          }
          if (answer === "optionTwo") {
            state.questions[qid].optionTwo.votes.push(authedUser);
          }
        }
      });
  },
});

export const { questionVote, updateQuestionVote, login, logout, updateUserNewQuestion } =
  appSlice.actions;

export default appSlice.reducer;

/* Async below here */

// Thunk for fetching all questions
export const fetchQuestions = createAsyncThunk("questions/fetchQuestions", async () => {
  const response = await _getQuestions();
  console.log("_getQuestions() API response: ", response);
  return response;
});

// Thunk for saving a new question
export const saveNewQuestion = createAsyncThunk("questions/saveNewQuestion", async (question) => {
  const response = await _saveQuestion(question);
  console.log("_saveQuestion() API response: ", response);
  return response;
});

// Thunk for fetching all user
export const fetchUsers = createAsyncThunk("users/fetchUsers", async () => {
  const response = await _getUsers();
  console.log("_getUsers() API response: ", response);
  return response;
});

// Thunk for saving user answer
export const saveUserAnswer = createAsyncThunk("users/saveUserAnswer", async (payload) => {
  const { authedUser, qid, answer } = payload;
  const response = await _saveQuestionAnswer({ authedUser, qid, answer });
  console.log("_saveQuestionAnswer() API response: ", response);
  if (response) {
    return payload;
  }
  return null;
});

/*  Selectors below here */
export const getAuthedUser = (state) => state.app.authedUser;
export const authStatus = (state) => state.app.authStatus;

export const getStatusUsers = (state) => state.app.statusUsersAPI;
export const getStatusQuestions = (state) => state.app.statusQuestionsAPI;

export const getAnswers = (state) => {
  const authedUser = getAuthedUser(state);
  const authedUserObject = state.app.users[authedUser];
  return authedUserObject.answers;
};

export const selectAllQuestions = (state) => state.app.questions;

export const allQuestionsAsArray = (state) => {
  const questions = selectAllQuestions(state);
  return Object.keys(questions).map((key) => questions[key]);
};

export const getAnsweredQuestionsAsArray = (state) => {
  const authedUser = getAuthedUser(state);
  const authedUserObject = state.app.users[authedUser];
  return Object.keys(authedUserObject.answers);
};

export const userInfo = (state) => {
  const authedUser = state.app.authedUser;
  const users = state.app.users;
  const user = users[authedUser];
  if (user) {
    const userName = user.name;
    const avatarURL = user.avatarURL;
    return { userName, avatarURL };
  }
  return { userName: "", avatarURL: "" };
};

export const userStatsArraySorted = (state) => {
  const users = state.app.users;
  let userStats = Object.keys(users).map((key) => {
    const questions = users[key].questions;
    const answers = users[key].answers;
    return {
      user: users[key].id,
      questions: questions.length,
      answers: Object.keys(answers).length,
      score: questions.length + Object.keys(answers).length,
    };
  });
  userStats = userStats.sort((a, b) => b.score - a.score);
  return userStats;
};
