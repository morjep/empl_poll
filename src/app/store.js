import { configureStore } from "@reduxjs/toolkit";
import questionsReducer from "./questionsSlice";
import usersReducer from "./usersSlice";
import appReducer from "./appSlice";

export const store = configureStore({
  reducer: {
    users: usersReducer,
    app: appReducer,
  },
});
