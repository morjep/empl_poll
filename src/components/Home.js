import { useSelector } from "react-redux";
import { Link } from "react-router-dom";
import {
  Avatar,
  Box,
  Container,
  Flex,
  FormControl,
  FormLabel,
  Input,
  Stack,
  Button,
  Heading,
  Text,
  useColorModeValue,
} from "@chakra-ui/react";
import { getStatusQuestions, allQuestionsAsArray } from "../app/appSlice";
import { getAnsweredQuestionsAsArray } from "../app/appSlice";

import styles from "./questions.module.css";

export function formatDate(timestamp) {
  const d = new Date(timestamp);
  const time = d.toLocaleTimeString("en-US");
  return time.substring(0, 5) + time.slice(-2) + " | " + d.toLocaleDateString();
}

// TODO: Check rubric wrt ordering of cards
const CardContainer = ({ heading, questions }) => {
  return (
    <Box rounded={"lg"} bg={"white"} boxShadow={"lg"} width={"4xl"} p={8} m={4}>
      <Heading fontSize={"4xl"} color={"teal.400"} borderBottom="1px" borderColor="blue.200">
        {heading}
      </Heading>
      <Flex flexWrap={"wrap"} mt={4}>
        {questions.map((question) => (
          <Link to={`/question/${question.id}`} key={question.id}>
            <Box
              m={15}
              p={10}
              rounded={"lg"}
              bg={"white"}
              boxShadow={"2xl"}
              _hover={{
                bg: "blue.200",
                boxShadow: "dark-lg",
              }}
              bgGradient="linear(to-br, green.200, teal.200)"
            >
              <Heading fontSize={"lg"}>{question.author}</Heading>
              <Text>{formatDate(question.timestamp)}</Text>
            </Box>
          </Link>
        ))}
      </Flex>
    </Box>
  );
};

export const Home = () => {
  const questionStatus = useSelector(getStatusQuestions);
  const allQuestionsArray = useSelector(allQuestionsAsArray);
  const answeredQuestionsArray = useSelector(getAnsweredQuestionsAsArray);

  if (questionStatus === "loading") {
    return <div>Loading</div>;
  }

  if (questionStatus === "failed") {
    return <div>Oops - something unexpected happened. Please try again.</div>;
  }

  const answeredQuestions = allQuestionsArray.filter((question) =>
    answeredQuestionsArray.includes(question.id)
  );

  const newQuestions = allQuestionsArray.filter(
    (question) => !answeredQuestionsArray.includes(question.id)
  );

  return (
    <Flex
      direction="column"
      minH={"100vh"}
      align={"center"}
      justify={"center"}
      bgGradient="linear(to-br, blue.200, blue.600)"
    >
      <CardContainer heading="New Questions" questions={newQuestions} />
      <CardContainer heading="Answered Questions" questions={answeredQuestions} />
    </Flex>
  );
};
