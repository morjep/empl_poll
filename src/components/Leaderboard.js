import { useSelector } from "react-redux";
import { userStatsArraySorted } from "../app/appSlice";
import {
  Flex,
  Heading,
  TableContainer,
  Table,
  TableCaption,
  Thead,
  Tbody,
  Tr,
  Th,
  Td,
  Tfoot,
} from "@chakra-ui/react";

export const Leaderboard = () => {
  const usersStats = useSelector(userStatsArraySorted);

  console.log("usersStats: ", usersStats);
  return (
    <Flex direction="column" align="center">
      <Heading color="blue.400" mt={25} mb={5}>
        Leaderboard
      </Heading>
      <Flex minWidth="max-content" maxWidth={800}>
        <TableContainer>
          <Table variant="striped" colorScheme="blue" size="md">
            <TableCaption>The poll score board</TableCaption>
            <Thead>
              <Tr>
                <Th>User</Th>
                <Th>Questions</Th>
                <Th>Answers</Th>
                <Th>Score</Th>
              </Tr>
            </Thead>
            <Tbody>
              {usersStats.map((user) => (
                <Tr key={user.user}>
                  <Td>{user.user}</Td>
                  <Td>{user.questions}</Td>
                  <Td>{user.answers}</Td>
                  <Td>{user.score}</Td>
                </Tr>
              ))}
            </Tbody>
          </Table>
        </TableContainer>
      </Flex>
    </Flex>
  );
};
export default Leaderboard;
