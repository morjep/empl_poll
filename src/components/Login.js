import { useDispatch } from "react-redux";
import { useState } from "react";
import {
  Flex,
  Box,
  FormControl,
  FormLabel,
  Input,
  Stack,
  Button,
  Heading,
  useColorModeValue,
} from "@chakra-ui/react";

import { login } from "../app/appSlice";

export const Login = () => {
  const dispatch = useDispatch();
  const [username, setUsername] = useState("");
  const [password, setPassword] = useState("");

  const handleSubmit = (e) => {
    e.preventDefault();
    dispatch(login({ name: username, pwd: password }));
    setUsername("");
    setPassword("");
  };

  const handleUsernameChange = (e) => {
    setUsername(e.target.value);
  };

  const handlePasswordChange = (e) => {
    setPassword(e.target.value);
  };

  return (
    <Flex
      minH={"100vh"}
      align={"center"}
      justify={"center"}
      bg={useColorModeValue("gray.50", "gray.800")}
    >
      <Stack spacing={8} mx={"auto"} maxW={"lg"} py={12} px={6}>
        <Stack align={"center"}>
          <Heading fontSize={"4xl"}>Login to your account</Heading>
        </Stack>
        <Box rounded={"lg"} bg={useColorModeValue("white", "gray.700")} boxShadow={"lg"} p={8}>
          <form onSubmit={handleSubmit}>
            <Stack spacing={4}>
              <FormControl id="username" isRequired>
                <FormLabel>Username</FormLabel>
                <Input type="text" onChange={handleUsernameChange} value={username} />
              </FormControl>
              <FormControl id="password" isRequired>
                <FormLabel>Password</FormLabel>
                <Input type="password" onChange={handlePasswordChange} value={password} />
              </FormControl>
              <Button
                type="submit"
                bg={"blue.400"}
                color={"white"}
                _hover={{
                  bg: "blue.600",
                }}
              >
                Login
              </Button>
            </Stack>
          </form>
        </Box>
      </Stack>
    </Flex>
  );
};
