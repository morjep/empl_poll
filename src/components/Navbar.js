import { Link } from "react-router-dom";
import { Flex, Text, Box, Spacer, useColorModeValue, Button, Avatar } from "@chakra-ui/react";
import { Stack, HStack, VStack } from "@chakra-ui/react";
import { userInfo, logout } from "../app/appSlice";
import { useSelector, useDispatch } from "react-redux";
import { buildQueries } from "@testing-library/react";

const links = [
  { name: "Home", href: "/" },
  { name: "Leaderboard", href: "/leaderboard" },
  { name: "New", href: "/add" },
];

const LinkTo = ({ name, link }) => {
  return (
    <Link to={link}>
      <Text
        fontSize="lg"
        color={"blue.400"}
        _hover={{
          color: "blue.600",
          borderBottom: "1px",
          borderColor: "blue.600", // fontWeight: "bold",
        }}
      >
        {name}
      </Text>
    </Link>
  );
};

export const Navbar = () => {
  const { userName, avatarURL } = useSelector(userInfo);
  const dispatch = useDispatch();

  return (
    <Flex as="nav" borderBottom="1px" borderColor="blue.200">
      <Box m={4}>
        <HStack spacing="24px">
          {links.map((link) => (
            <LinkTo key={link.name} name={link.name} link={link.href} />
          ))}
        </HStack>
      </Box>
      <Spacer />
      <Box m={4}>
        <HStack spacing="2px">
          <Avatar size={"md"} src={avatarURL} />
          <Text color={"blue.600"} fontWeight={"bolder"} fontSize="lg" pl={5}>
            {userName}
          </Text>
        </HStack>
      </Box>
      <Spacer />
      <Box m={4}>
        <Link to="/" onClick={() => dispatch(logout())}>
          <Button
            type="submit"
            bg={"blue.400"}
            color={"white"}
            _hover={{
              bg: "blue.600",
            }}
          >
            Logout
          </Button>
        </Link>
      </Box>
    </Flex>
  );
};
export default Navbar;
