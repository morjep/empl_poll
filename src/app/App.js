import { useEffect } from "react";
import { useSelector, useDispatch } from "react-redux";
import { Routes, Route } from "react-router-dom";
import { ChakraProvider, theme } from "@chakra-ui/react";
import { Flex, Heading } from "@chakra-ui/react";

import { Home } from "../components/Home";
import { NewQuestion } from "../components/NewQuestion";
import { Leaderboard } from "../components/Leaderboard";
import { Question } from "../components/Question";
import { getStatusUsers, fetchUsers, getAuthedUser } from "./appSlice";
import { getStatusQuestions, fetchQuestions } from "./appSlice";

import { Navbar } from "../components/Navbar";
import { Login } from "../components/Login";

const NoMatch = () => (
  <Flex justify={"center"} m={15}>
    <Heading>404 - Page not found!</Heading>
  </Flex>
);

function App() {
  const dispatch = useDispatch();
  const usersStatus = useSelector(getStatusUsers);
  const questionStatus = useSelector(getStatusQuestions);
  const authedUser = useSelector(getAuthedUser);

  useEffect(() => {
    if (usersStatus === "idle") {
      dispatch(fetchUsers());
    }
  }, [usersStatus, dispatch]);

  useEffect(() => {
    if (questionStatus === "idle") {
      dispatch(fetchQuestions());
    }
  }, [questionStatus, dispatch]);

  return (
    <ChakraProvider theme={theme}>
      <div className="App">
        {authedUser && <Navbar />}
        {authedUser === null ? (
          <Login />
        ) : (
          <Routes>
            <Route exact path="/" element={<Home />} />
            <Route path="/add" element={<NewQuestion />} />
            <Route path="/leaderboard" element={<Leaderboard />} />
            <Route path="/question/:id" element={<Question />} />
            <Route path="*" element={<NoMatch />} />
          </Routes>
        )}
      </div>
    </ChakraProvider>
  );
}

export default App;
