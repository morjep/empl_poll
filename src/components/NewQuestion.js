import { useState } from "react";
import { useSelector, useDispatch } from "react-redux";
import { saveNewQuestion } from "../app/appSlice";
import { authedUser, userInfo } from "../app/appSlice";
import { useNavigate } from "react-router-dom";
import {
  Avatar,
  Flex,
  FormControl,
  FormLabel,
  Input,
  Stack,
  Button,
  Heading,
  useColorModeValue,
} from "@chakra-ui/react";

export const NewQuestion = () => {
  const dispatch = useDispatch();
  const user = useSelector(authedUser);
  const { userName, avatarURL } = useSelector(userInfo);

  let navigate = useNavigate();

  const [optionOne, setOptionOne] = useState("");
  const [optionTwo, setOptionTwo] = useState("");

  const handleSubmit = (e) => {
    e.preventDefault();
    const author = user;
    const question = {
      optionOneText: optionOne,
      optionTwoText: optionTwo,
      author: author,
    };
    dispatch(saveNewQuestion(question));
    setOptionOne("");
    setOptionTwo("");
    navigate("/");
  };

  const handleOptionOneChange = (e) => {
    setOptionOne(e.target.value);
  };

  const handleOptionTwoChange = (e) => {
    setOptionTwo(e.target.value);
  };

  return (
    <Flex
      direction="row"
      minH={"100vh"}
      align={"center"}
      justify={"center"}
      bgGradient="linear(to-br, blue.200, blue.600)"
    >
      <Flex direction="column" boxSize={"md"} rounded={"lg"} justify={"top"} align="center">
        <Heading color="blue.100" mb={5} size={"2xl"}>
          Create new poll
        </Heading>
        <Avatar size={"2xl"} src={avatarURL} mb={10} />
      </Flex>
      <Flex
        boxSize={"md"}
        rounded={"lg"}
        bg={useColorModeValue("white", "gray.700")}
        boxShadow={"lg"}
        p={8}
        justify={"center"}
      >
        <form onSubmit={handleSubmit}>
          <Stack spacing={8}>
            <Heading fontSize={"4xl"} color={"teal.400"}>
              Would you rather....?
            </Heading>

            <FormControl id="optionOne" isRequired>
              <FormLabel>First option</FormLabel>
              <Input type="text" onChange={handleOptionOneChange} value={optionOne} />
            </FormControl>
            <FormControl id="optionTwo" isRequired>
              <FormLabel>Second option</FormLabel>
              <Input type="text" onChange={handleOptionTwoChange} value={optionTwo} />
            </FormControl>
            <Button
              type="submit"
              bg={"teal.400"}
              color={"white"}
              _hover={{
                bg: "teal.600",
              }}
            >
              Submit
            </Button>
          </Stack>
        </form>
      </Flex>
    </Flex>
  );
};
export default NewQuestion;
