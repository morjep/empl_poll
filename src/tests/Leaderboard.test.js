import { render } from "@testing-library/react";
import { Provider } from "react-redux";
import { Leaderboard } from "../components/Leaderboard";
import { ChakraProvider, theme } from "@chakra-ui/react";
import { store } from "../app/store";

describe("Leaderboard", () => {
  it("should render correctly", () => {
    const { container } = render(
      <Provider store={store}>
        <ChakraProvider theme={theme}>
          <Leaderboard />
        </ChakraProvider>
      </Provider>
    );
    expect(container).toMatchSnapshot();
  });
});
