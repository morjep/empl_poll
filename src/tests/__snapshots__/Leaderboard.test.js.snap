// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Leaderboard should render correctly 1`] = `
<div>
  <div
    class="css-4ur6rm"
  >
    <h2
      class="chakra-heading css-1bmd4e0"
    >
      Leaderboard
    </h2>
    <div
      class="css-1uv0lsc"
    >
      <div
        class="chakra-table__container css-zipzvv"
      >
        <table
          class="chakra-table css-5605sr"
          role="table"
        >
          <caption
            class="css-7478fk"
          >
            The poll score board
          </caption>
          <thead
            class="css-0"
          >
            <tr
              class="css-0"
              role="row"
            >
              <th
                class="css-pndeee"
              >
                User
              </th>
              <th
                class="css-1i0h9sv"
              >
                Questions
              </th>
              <th
                class="css-1i0h9sv"
              >
                Answers
              </th>
              <th
                class="css-qbq42q"
              >
                Score
              </th>
            </tr>
          </thead>
          <tbody
            class="css-0"
          />
        </table>
      </div>
    </div>
  </div>
  <span />
</div>
`;
